### PR TITLE
L2-944: Do not allow commitments above max

### PR DESCRIPTION
**Pull request by lmuntaner**
# Motivation

User should not be able to commit if the amount is over the total maximum.

# Changes

* Add one more check in "validParticipation" project util.

# Tests

* Add test for new case.

### CODE REVIEW

**Comment by lmuntaner** [comment]
UX:

Trying to increase commitment by 30 when there is only 21 left to reach maximum.

https://user-images.githubusercontent.com/4550653/184851984-4829f8d9-3dab-4f48-a302-9ee447433604.mov

**Comment by lmuntaner** [comment]
@peterpeterparker @mstrasinskis please review

File: frontend/src/lib/utils/projects.utils.ts [hunk 1]
```diff
@@ -132,6 +132,13 @@ const commitmentTooLarge = ({
   summary: SnsSummary;
   amountE8s: bigint;
 }): boolean => summary.swap.init.max_participant_icp_e8s < amountE8s;
+const commitmentExceedsAmountLeft = ({
```
Comment (peterpeterparker) [inline frontend/src/lib/utils/projects.utils.ts:135]: what you think, makes sense to add this new checks to `canUserParticipateToSwap` too?

File: frontend/src/lib/utils/projects.utils.ts [hunk 2]
```diff
@@ -214,6 +221,25 @@ export const validParticipation = ({
...
+      summary: project.summary,
+      amountE8s: totalCommitment,
+    })
+  ) {
+    return {
+      valid: false,
+      labelKey: "error__sns.commitment_exceeds_current_allowed",
+      substitutions: {
+        $commitment: formatICP({ value: totalCommitment }),
+        $maxCommitment: formatICP({
```
Comment (peterpeterparker) [inline frontend/src/lib/utils/projects.utils.ts:235]: should we rename the variable `maxCommitment` to `remainingCommitment`?

can maybe make the label easier to translate if we ever translate it

File: frontend/src/tests/lib/utils/projects.utils.spec.ts [hunk 3]
```diff
@@ -454,6 +458,34 @@ describe("project-utils", () => {
...
+              max_participant_icp_e8s: maxE8s,
+            },
+          },
+        },
+      };
+      const { valid } = validParticipation({
+        project,
+        amount: ICP.fromE8s(participationE8s),
+      });
+      expect(valid).toBe(false);
```
Comment (peterpeterparker) [inline frontend/src/tests/lib/utils/projects.utils.spec.ts:487]: is there also a test that checks that it returns `true` when `commitmentExceedsAmountLeft` is ok?

File: frontend/src/lib/utils/projects.utils.ts (hunk 2 above)
Comment (lmuntaner) [reply frontend/src/lib/utils/projects.utils.ts:235]: makes sense

File: frontend/src/tests/lib/utils/projects.utils.spec.ts (hunk 3 above)
Comment (lmuntaner) [reply frontend/src/tests/lib/utils/projects.utils.spec.ts:487]: yes, the valid participation is a participation that doesn't exceed the amount left.

File: frontend/src/lib/utils/projects.utils.ts (hunk 1 above)
Comment (lmuntaner) [reply frontend/src/lib/utils/projects.utils.ts:135]: no, this only happens when we have some amount that the user wants to participate with.

`canUserParticipateToSwap` doesn't have any extra amount. It's more, can the user participate with just 1ICP?

File: frontend/src/lib/utils/projects.utils.ts (hunk 2 above)
Comment (lmuntaner) [reply frontend/src/lib/utils/projects.utils.ts]: done

**Comment by lmuntaner** [comment]
@peterpeterparker ready for another review

File: frontend/src/lib/utils/projects.utils.ts (hunk 1 above)
Comment (peterpeterparker) [reply frontend/src/lib/utils/projects.utils.ts:135]: thanks for the explanation. so, I was a bit confuse. can we make the name of the function more obvious or add a comment?

File: frontend/src/tests/lib/utils/projects.utils.spec.ts (hunk 3 above)
Comment (peterpeterparker) [reply frontend/src/tests/lib/utils/projects.utils.spec.ts:487]: super

File: frontend/src/lib/utils/projects.utils.ts (hunk 1 above)
Comment (lmuntaner) [reply frontend/src/lib/utils/projects.utils.ts:135]: There is a comment in the canUserParticipateToSwap:
```
/**
 * To participate to a swap:
 *
 * - the sale's lifecycle should be Open
 * - user commitment should not be bigger than the maximal commitment allowed per user
 * - the maximal commitment allowed per user should be defined
 */
```
Just not showed here.

Or do you mean a comment in `commitmentExceedsAmountLeft`?

Comment (peterpeterparker) [reply frontend/src/lib/utils/projects.utils.ts:135]: i mean `commitmentExceedsAmountLeft`

Comment (lmuntaner) [reply frontend/src/lib/utils/projects.utils.ts:138]: done

**Comment by lmuntaner** [comment]
@peterpeterparker ready for another review

**Review by peterpeterparker — APPROVED** [review]
thx!

**Pull request merged**